### PR TITLE
[WIP] threading utilities

### PR DESCRIPTION
**Pull request by fredrikekre**

### CODE REVIEW

File: src/threadingtools.jl [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+Create a `ThreadVector` with the same length as number of threads
+"""
+macro threaded(ex)
+    _threaded(ex)
+end
+
+function _threaded(ex)
+    t = nthreads()
+    if isa(ex, Symbol) # an identifier
+        return esc(quote
```
Comment (fredrikekre) [inline src/threadingtools.jl]: I need help with the macro hygiene here :p

**Comment by fredrikekre** [comment]
Meh:
```julia
julia> HyperElasticity.solve(40);
Solving: Time: 0:10:22 (7 iterations)
+--------------------------------+-----------+--------+---------+
| Wall time elapsed since start  |     622 s |        |         |
|                                |           |        |         |
| Section              | n calls | wall time | % tot  | % timed |
+----------------------+---------+-----------+--------+---------+
| linear solve         |       6 |     349 s |  56.1% |   56.1% |
| assemble             |       7 |     144 s |  23.1% |   23.1% |
| assemble threaded    |       7 |     120 s |  19.2% |   19.2% |
| create sparse mat... |       7 |    6.69 s |  1.08% |   1.08% |
| create sparse mat... |       7 |    2.86 s |  0.46% |   0.46% |
+----------------------+---------+-----------+--------+---------+
```

**Comment by KristofferC** [comment]
Tycker det är lite konstigt.

**Comment by fredrikekre** [comment]
yes

**Comment by fredrikekre** [comment]
For 10x10x10 elements it is about double as fast though:

```julia
julia> HyperElasticity.solve(10);
Solving: Time: 0:00:05 (5 iterations)
+--------------------------------+-----------+--------+---------+
| Wall time elapsed since start  |    4.67 s |        |         |
|                                |           |        |         |
| Section              | n calls | wall time | % tot  | % timed |
+----------------------+---------+-----------+--------+---------+
| assemble             |       5 |    2.46 s |  52.7% |   53.8% |
| assemble threaded    |       5 |    1.34 s |  28.6% |   29.2% |
| linear solve         |       4 |     0.7 s |    15% |   15.3% |
| create sparse mat... |       5 |  0.0446 s | 0.956% |  0.977% |
| create sparse mat... |       5 |  0.0319 s | 0.682% |  0.697% |
+----------------------+---------+-----------+--------+---------+
```

**Comment by fredrikekre** [comment]
```julia
julia> HyperElasticity.solve(50)
+--------------------------------+-----------+--------+---------+
| Wall time elapsed since start  |     151 s |        |         |
|                                |           |        |         |
| Section              | n calls | wall time | % tot  | % timed |
+----------------------+---------+-----------+--------+---------+
| assemble             |       1 |    41.2 s |  27.2% |   27.3% | # normal assembling
| assemble threaded    |       1 |    35.8 s |  23.6% |   23.7% | # just a normal @threads loop
| ass. th. loop sh     |       1 |    34.5 s |  22.8% |   22.8% | # in a function with sizehint
| ass. th. loop        |       1 |    34.3 s |  22.7% |   22.7% | # in a function
+----------------------+---------+-----------+--------+---------+
```

**Pull request closed without merging**